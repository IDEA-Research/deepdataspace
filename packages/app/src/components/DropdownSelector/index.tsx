import React from 'react';
import { Button, Space, Checkbox, Dropdown, Radio } from 'antd';
import { DownOutlined } from '@ant-design/icons';
import { globalLocaleText } from 'dds-utils/locale';
import styles from './index.less';

export type ValueType = string | string[] | number | number[];

export interface IProps<T, P> {
  data?: T[];
  value?: P;
  multiple?: boolean;
  filterOptionValue?: (option: any) => P;
  filterOptionName?: (option: any) => string | number;
  onChange?: (value: P) => void;
  type?: 'link' | 'default' | 'text' | 'ghost' | 'primary' | 'dashed';
  ghost?: boolean;
  className?: string;
  children?: React.ReactNode;
  customOverlay?: React.ReactNode;
}

const DropdownSelector: <Option = any, Value extends ValueType = ValueType>(
  props: IProps<Option, Value>,
) => React.ReactElement = (props) => {
  const {
    data,
    multiple,
    type = 'primary',
    ghost = true,
    value,
    filterOptionValue,
    filterOptionName,
    onChange,
    className,
    children,
    customOverlay,
  } = props;

  const TargetElement = multiple ? Checkbox : Radio;
  const onChangeValue = (e: any) => {
    if (!onChange) return;
    if (multiple) {
      onChange(e);
    } else {
      onChange(e.target.value);
    }
  };

  return (
    <Dropdown
      overlayClassName={styles.dropdownSelector}
      trigger={['click']}
      dropdownRender={() => (
        <div className={styles.dropdownWrap}>
          {customOverlay ? (
            customOverlay
          ) : (
            // @ts-ignore
            <TargetElement.Group
              className={styles.dropdownBox}
              onChange={onChangeValue}
              value={value}
            >
              <Space direction="vertical">
                {data?.map((item, index) => {
                  const value = filterOptionValue
                    ? filterOptionValue(item)
                    : (item as string | number);
                  const name = filterOptionName
                    ? filterOptionName(item)
                    : (item as string | number);
                  return (
                    <TargetElement key={index} value={value}>
                      {globalLocaleText(name)}
                    </TargetElement>
                  );
                })}
              </Space>
            </TargetElement.Group>
          )}
        </div>
      )}
    >
      <Button className={className} type={type} ghost={ghost}>
        {children}
        <DownOutlined />
      </Button>
    </Dropdown>
  );
};

export default DropdownSelector;
