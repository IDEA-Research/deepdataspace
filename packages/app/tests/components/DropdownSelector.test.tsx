import { umiRender } from '../test-utils';
import DropdownSelector from '@/components/DropdownSelector';

test('render DropdownSelector by snapshot', () => {
  const { container } = umiRender(<DropdownSelector />);
  expect(container).toMatchSnapshot();
});
