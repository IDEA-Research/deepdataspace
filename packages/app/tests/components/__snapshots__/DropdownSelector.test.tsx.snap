// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render DropdownSelector by snapshot 1`] = `
<div>
  <button
    class="ant-btn css-dev-only-do-not-override-190m0jy ant-btn-primary ant-btn-background-ghost ant-dropdown-trigger"
    type="button"
  >
    <span
      aria-label="down"
      class="anticon anticon-down"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="down"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
        />
      </svg>
    </span>
  </button>
</div>
`;
